hardware_interface/test/test_components/test_system.cpp

```diff
@@ -78,9 +131,22 @@ class TestSystem : public SystemInterface
     return command_interfaces;
   }
 
-  return_type read() override { return return_type::OK; }
+  return_type read(std::vector<hardware_interface::Variant> & states) override
+  {
+    // get some states from robots
+    states[0].set_value(false);    // there is still some implicit knowledge about order of interfaces (the same order as they info is exported)
+    states[1].set_value(1.2);
+
+    return return_type::OK;
```

Comment:
Because of how branch prediction works in modern CPUs if statements that nearly always go one way (in this case it will nearly always be OK) should be virtually no performance cost to checking on every iteration.

Reply:
Typically, this function would fail if communication to hardware is broken.